File: src/test/java/com/zepben/vertxutils/routing/handlers/PathParamsHandlerTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.handlers;

import com.zepben.vertxutils.routing.ErrorFormatter;
import com.zepben.vertxutils.routing.RoutingContextEx;
import com.zepben.vertxutils.routing.handlers.params.*;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.ext.web.RoutingContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.util.Arrays;
import java.util.Collections;

import static java.util.stream.Collectors.toList;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.*;

public class PathParamsHandlerTest {

    private final PathParamRule<Integer> numParam = PathParamRule.Companion.of("num", ParamType.INSTANCE.getINT());
    private final PathParamRule<Integer> num2Param = PathParamRule.Companion.of("num2", ParamType.INSTANCE.getINT());

    private final ArgumentCaptor<PathParams> paramsCaptor = ArgumentCaptor.forClass(PathParams.class);
    private final RoutingContext context = mock(RoutingContext.class);
    private final HttpServerResponse response = mock(HttpServerResponse.class, RETURNS_SELF);

    @BeforeEach
    public void setUp() {
        doReturn(response).when(context).response();
    }

    @Test
    public void callContextNext() {
        new PathParamsHandler().handle(context);
        verify(context).next();
    }

    @Test
    public void pathParam() {
        PathParamsHandler handler = new PathParamsHandler(numParam);

        doReturn("4").when(context).pathParam(numParam.getName());
        handler.handle(context);

        verify(context).put(eq(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY()), paramsCaptor.capture());
        PathParams params = paramsCaptor.getValue();
        assertThat(params.get(numParam), is(4));
    }

    @Test
    public void getFromContext() {
        PathParams expected = new PathParams(Collections.emptyMap());
        doReturn(expected).when(context).get(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY());
        assertThat(RoutingContextEx.INSTANCE.getPathParams(context), is(expected));
    }

    @Test
    public void pathParamMissing() {
        PathParamsHandler handler = new PathParamsHandler(numParam);
        handler.handle(context);

        verifyBadParamResponse(BadParamException.Companion.missingParam(numParam.getName()));
    }

    @Test
    public void pathParamBad() {
        PathParamsHandler handler = new PathParamsHandler(numParam, num2Param);
        doReturn("not a number").when(context).pathParam(numParam.getName());
        doReturn("true").when(context).pathParam(num2Param.getName());

        ValueConversionException ex1 = captureException(() -> numParam.getConverter().convert("not a number"), ValueConversionException.class);
        ValueConversionException ex2 = captureException(() -> num2Param.getConverter().convert("true"), ValueConversionException.class);
        handler.handle(context);

        verifyBadParamResponse(
            BadParamException.Companion.invalidParam(numParam, "not a number", ex1.getMessage()),
            BadParamException.Companion.invalidParam(num2Param, "true", ex2.getMessage()));
    }

    @SuppressWarnings("SameParameterValue")
    private <T extends Exception> T captureException(Runnable runnable, Class<T> expectedExType) {
        try {
            runnable.run();
        } catch (Exception ex) {
            return expectedExType.cast(ex);
        }

        throw new AssertionError("Expected exception but none was thrown");
    }

    private void verifyBadParamResponse(BadParamException... e) {
        verify(response).setStatusCode(400);
        String json = ErrorFormatter.INSTANCE.asJson(Arrays.stream(e).map(Throwable::getMessage).collect(toList()));
        verify(response).end(json);
        verify(context, never()).next();
    }
}

File: src/test/java/com/zepben/vertxutils/routing/handlers/PathParamsHandlerTest.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing.handlers

import com.zepben.testutils.junit.SystemLogExtension
import com.zepben.vertxutils.routing.ErrorFormatter.asJson
import com.zepben.vertxutils.routing.RoutingContextEx.PATH_PARAMS_KEY
import com.zepben.vertxutils.routing.RoutingContextEx.getPathParams
import com.zepben.vertxutils.routing.handlers.params.*
import io.vertx.core.http.HttpServerResponse
import io.vertx.ext.web.RoutingContext
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.sameInstance
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.*

class PathParamsHandlerTest {

    companion object {

        @JvmField
        @RegisterExtension
        val systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess()

    }

    private val numParam = PathParamRule.of("num", ParamType.INT)
    private val num2Param = PathParamRule.of("num2", ParamType.INT)

    private val paramsCaptor = ArgumentCaptor.forClass(PathParams::class.java)
    private val response = mock<HttpServerResponse>(RETURNS_SELF)
    private val context = mock<RoutingContext>().also {
        doReturn(response).`when`(it).response()
    }

    @Test
    fun callContextNext() {
        PathParamsHandler().handle(context)
        verify(context).next()
    }

    @Test
    fun pathParam() {
        val handler = PathParamsHandler(numParam)

        doReturn("4").`when`(context).pathParam(numParam.name)
        handler.handle(context)

        verify(context).put(eq(PATH_PARAMS_KEY), paramsCaptor.capture())
        val params = paramsCaptor.getValue()
        assertThat(params[numParam], equalTo(4))
    }

    @Test
    fun getFromContext() {
        val expected = PathParams(emptyMap())
        doReturn(expected).`when`(context).get<Any?>(PATH_PARAMS_KEY)
        assertThat(getPathParams(context), sameInstance(expected))
    }

    @Test
    fun pathParamMissing() {
        val handler = PathParamsHandler(numParam)
        handler.handle(context)

        verifyBadParamResponse(BadParamException.missingParam(numParam.name))
    }

    @Test
    fun pathParamBad() {
        val handler = PathParamsHandler(numParam, num2Param)
        doReturn("not a number").`when`(context).pathParam(numParam.name)
        doReturn("true").`when`(context).pathParam(num2Param.name)

        val ex1 = captureException<ValueConversionException> { numParam.converter.convert("not a number") }
        val ex2 = captureException<ValueConversionException> { num2Param.converter.convert("true") }
        handler.handle(context)

        verifyBadParamResponse(
            BadParamException.invalidParam(numParam, "not a number", ex1.message),
            BadParamException.invalidParam(num2Param, "true", ex2.message),
        )
    }

    private inline fun <reified T : Exception> captureException(runnable: Runnable): T {
        try {
            runnable.run()
        } catch (ex: Exception) {
            return ex as T
        }

        throw AssertionError("Expected exception but none was thrown")
    }

    private fun verifyBadParamResponse(vararg e: BadParamException) {
        verify(response).statusCode = 400
        val json = asJson(e.map { it.message })
        verify(response).end(json)
        verify(context, never()).next()
    }

}
